sbncode/TPCReco/VertexStub/StubBuilder.cxx

```diff
@@ -0,0 +1,232 @@
+#include "StubBuilder.h"
+
+// Helper functions
+bool HitOnTrack(const art::Ptr<recob::Hit> &hit,
+    const art::Ptr<recob::Track> &trk,
+    const std::vector<art::Ptr<recob::Hit>> &trk_hits,
+    const std::vector<const recob::TrackHitMeta *> &trk_thms) {
+
+  if (!trk) return false; // no track: hit can't be on a track
+    
+  for (unsigned i = 0; i < trk_hits.size(); i++) {
+    // Found the hit on the track!
+    if (hit == trk_hits[i]) { 
+      // Is the hit part of the track Calo?
+      if (trk_thms[i]->Index() != std::numeric_limits<int>::max() && trk->HasValidPoint(trk_thms[i]->Index())) { 
+        return true;
+      }
+      break;
+    }
+  }
+  return false;
+}
+
+int FindPFPInd(
+  const std::vector<art::Ptr<recob::Hit>> &hits,
+  const std::vector<art::Ptr<recob::PFParticle>> &pfps,
+  const std::vector<std::vector<art::Ptr<recob::Hit>>> &pfp_hits) {
+
+  std::vector<unsigned> pfp_nhit(pfps.size(), 0);
+
+  for (const art::Ptr<recob::Hit> h: hits) {
+    for (unsigned i_pfp = 0; i_pfp < pfps.size(); i_pfp++) {
+      const std::vector<art::Ptr<recob::Hit>> &thispfpHits = pfp_hits[i_pfp];
+      bool found = false;
+      for (const art::Ptr<recob::Hit> &pfp_h: thispfpHits) {
+        if (h == pfp_h) {
+          found = true;
+          pfp_nhit[i_pfp] ++;
+          break;
+        }
+      }
+      if (found) break;
+    }
+  }
+
+  for (unsigned i_pfp = 0; i_pfp < pfps.size(); i_pfp++) {
+    if (pfp_nhit[i_pfp] >= hits.size() / 2) return i_pfp;
+  }
+
+  return -1;
+}
+
+std::vector<art::Ptr<recob::Hit>> CollectHits(
+    const std::vector<art::Ptr<recob::Hit>> &hits, 
+    const sbn::VertexHit &vhit, 
+    const recob::Hit &vhit_hit, 
+    const geo::Point_t &vertex, 
+    const geo::GeometryCore *geo, 
+    const detinfo::DetectorPropertiesData &dprop) {
+
+  // project the vertex onto the wireplane
+  float vert_wf = geo->WireCoordinate(vertex, vhit_hit.WireID());
+  float vert_x = vertex.x();
+
+  // get the vertex hit-coordinates
+  int hit_w = vhit_hit.WireID().Wire;
+  float hit_x = dprop.ConvertTicksToX(vhit_hit.PeakTime(), vhit_hit.WireID());
+  geo::PlaneID vhit_plane = vhit_hit.WireID();
+
+  // get the line-segment slope / intercept
+  float slope = (hit_x - vert_x) / (hit_w - vert_wf);
```

Comment:
I don't know if that's a good idea. 

I think that it may happen sometimes that the hit wire is the same as the vertex wire, and the code should not crash in that case.

Reply:
What should it do in that case?

Reply:
Ok -- then only hits on the (shared) vertex + hit-end-point wire should be included.

I have added in logic to do this.